File: benchmark/json_codec.py
import sys
import time

# Faithful port of benchmark/json_codec.dart's guest program — same
# recursive-descent JSON parser and serializer over generated docs.

class Parser:
    __slots__ = ('s', 'pos')

    def __init__(self, s):
        self.s = s
        self.pos = 0

    def ws(self):
        s = self.s
        n = len(s)
        while self.pos < n:
            c = ord(s[self.pos])
            if c == 32 or c == 9 or c == 10 or c == 13:
                self.pos += 1
            else:
                break

    def value(self):
        self.ws()
        c = ord(self.s[self.pos])
        if c == 123:
            return self.object()
        if c == 91:
            return self.array()
        if c == 34:
            return self.string()
        if c == 116:
            self.pos += 4
            return True
        if c == 102:
            self.pos += 5
            return False
        if c == 110:
            self.pos += 4
            return None
        return self.number()

    def object(self):
        self.pos += 1
        m = {}
        self.ws()
        if ord(self.s[self.pos]) == 125:
            self.pos += 1
            return m
        while True:
            self.ws()
            key = self.string()
            self.ws()
            self.pos += 1
            m[key] = self.value()
            self.ws()
            c = ord(self.s[self.pos])
            if c == 44:
                self.pos += 1
                continue
            self.pos += 1
            return m

    def array(self):
        self.pos += 1
        l = []
        self.ws()
        if ord(self.s[self.pos]) == 93:
            self.pos += 1
            return l
        while True:
            l.append(self.value())
            self.ws()
            c = ord(self.s[self.pos])
            if c == 44:
                self.pos += 1
                continue
            self.pos += 1
            return l

    def string(self):
        self.pos += 1
        start = self.pos
        buf = None
        s = self.s
        while self.pos < len(s):
            c = ord(s[self.pos])
            if c == 34:
                break
            if c == 92:
                if buf is None:
                    buf = []
                buf.append(s[start:self.pos])
                self.pos += 1
                e = ord(s[self.pos])
                if e == 110:
                    buf.append('\n')
                elif e == 116:
                    buf.append('\t')
                elif e == 117:
                    buf.append(chr(self.hex4()))
                else:
                    buf.append(chr(e))
                self.pos += 1
                start = self.pos
                continue
            self.pos += 1
        tail = s[start:self.pos]
        self.pos += 1
        if buf is None:
            return tail
        buf.append(tail)
        return ''.join(buf)

    def hex4(self):
        v = 0
        s = self.s
        for i in range(4):
            c = ord(s[self.pos])
            d = c - 48
            if d > 9:
                d = (c | 32) - 87
            v = v * 16 + d
            self.pos += 1
        return v

    def number(self):
        s = self.s
        neg = False
        if ord(s[self.pos]) == 45:
            neg = True
            self.pos += 1
        n = 0
        while self.pos < len(s):
            c = ord(s[self.pos])
            if c < 48 or c > 57:
                break
            n = n * 10 + c - 48
            self.pos += 1
        if self.pos < len(s) and ord(s[self.pos]) == 46:
            self.pos += 1
            scale = 0.1
            frac = 0.0
            while self.pos < len(s):
                c = ord(s[self.pos])
                if c < 48 or c > 57:
                    break
                frac += (c - 48) * scale
                scale *= 0.1
                self.pos += 1
            v = n + frac
            return -v if neg else v
        return -n if neg else n


def write_string(v, b):
    b.append('"')
    start = 0
    for i in range(len(v)):
        c = ord(v[i])
        if c == 34 or c == 92:
            if i > start:
                b.append(v[start:i])
            b.append('\\')
            b.append(chr(c))
            start = i + 1
        elif c == 10:
            if i > start:
                b.append(v[start:i])
            b.append('\\n')
            start = i + 1
    if len(v) > start:
        b.append(v[start:])
    b.append('"')


def write_value(v, b):
    if v is None:
        b.append('null')
        return
    if isinstance(v, bool):
        b.append('true' if v else 'false')
        return
    if isinstance(v, (int, float)):
        b.append(str(v))
        return
    if isinstance(v, str):
        write_string(v, b)
        return
    if isinstance(v, list):
        b.append('[')
        for i in range(len(v)):
            if i > 0:
                b.append(',')
            write_value(v[i], b)
        b.append(']')
        return
    b.append('{')
    first = True
    for key, val in v.items():
        if not first:
            b.append(',')
        first = False
        write_string(key, b)
        b.append(':')
        write_value(val, b)
    b.append('}')


def make_doc(seed):
    x = seed

    def nxt():
        nonlocal x
        x = (x * 1103515245 + 12345) & 0x7fffffff
        return x

    b = []
    b.append('{"items":[')
    for i in range(40):
        if i > 0:
            b.append(',')
        b.append('{"id":')
        b.append(str(nxt() % 100000))
        b.append(',"name":"user_')
        b.append(str(nxt() % 9999))
        b.append('","tags":["alpha","b')
        b.append(str(nxt() % 999))
        b.append('","g"],"score":')
        b.append(str((nxt() % 10000) / 100))
        b.append(',"ok":')
        b.append('true' if nxt() % 3 == 0 else 'false')
        b.append(',"meta":null}')
    b.append('],"count":40}')
    return ''.join(b)


def checksum(v):
    if isinstance(v, list):
        n = len(v)
        for e in v:
            n += checksum(e)
        return n
    if isinstance(v, dict):
        n = len(v)
        for k, e in v.items():
            n += len(k) + checksum(e)
        return n
    if isinstance(v, str):
        return len(v)
    if isinstance(v, bool):
        return 1
    if isinstance(v, int):
        return v & 0xff
    if isinstance(v, float):
        return int(v) & 0xff
    return 0 if v is None else 1


def main(n):
    total = 0
    for i in range(n):
        doc = make_doc(123456789 + i * 2654435761)
        parsed = Parser(doc).value()
        out = []
        write_value(parsed, out)
        total += len(''.join(out)) + checksum(parsed)
    return total


if __name__ == '__main__':
    iterations = int(sys.argv[1]) if len(sys.argv) > 1 else 4000
    samples = int(sys.argv[2]) if len(sys.argv) > 2 else 7
    for _ in range(2):
        main(50)
    times = []
    for _ in range(samples):
        t0 = time.perf_counter()
        result = main(iterations)
        times.append((time.perf_counter() - t0) * 1000)
    times.sort()
    median = times[len(times) // 2]
    print(f'json_codec median_ms={median:.3f} '
          f'min_ms={times[0]:.3f} '
          f'max_ms={times[-1]:.3f} '
          f'ns/iteration={median * 1000000 / iterations:.2f}')
    print(f'checksum={result}')

File: benchmark/json_codec.dart
import 'package:dart_eval/dart_eval.dart';

// A hand-rolled JSON parser and serializer — the string-scanning,
// map/list-building codec pattern real codecs share (no dart:convert).
// Run with: dart compile exe benchmark/json_codec.dart -o json_codec.exe
// Then: json_codec.exe [iterations] [samples]

const _source = r'''
class Parser {
  Parser(this.s);
  final String s;
  int pos = 0;

  int get len => s.length;

  void ws() {
    while (pos < len) {
      final c = s.codeUnitAt(pos);
      if (c == 32 || c == 9 || c == 10 || c == 13) { pos++; } else { break; }
    }
  }

  Object? value() {
    ws();
    final c = s.codeUnitAt(pos);
    if (c == 123) return object();
    if (c == 91) return array();
    if (c == 34) return string();
    if (c == 116) { pos += 4; return true; }
    if (c == 102) { pos += 5; return false; }
    if (c == 110) { pos += 4; return null; }
    return number();
  }

  Map<String, Object?> object() {
    pos++;
    final m = <String, Object?>{};
    ws();
    if (s.codeUnitAt(pos) == 125) { pos++; return m; }
    while (true) {
      ws();
      final key = string();
      ws();
      pos++;
      m[key] = value();
      ws();
      final c = s.codeUnitAt(pos);
      if (c == 44) { pos++; continue; }
      pos++;
      return m;
    }
  }

  List<Object?> array() {
    pos++;
    final l = <Object?>[];
    ws();
    if (s.codeUnitAt(pos) == 93) { pos++; return l; }
    while (true) {
      l.add(value());
      ws();
      final c = s.codeUnitAt(pos);
      if (c == 44) { pos++; continue; }
      pos++;
      return l;
    }
  }

  String string() {
    pos++;
    var start = pos;
    StringBuffer? buf;
    while (pos < len) {
      final c = s.codeUnitAt(pos);
      if (c == 34) break;
      if (c == 92) {
        (buf ??= StringBuffer()).write(s.substring(start, pos));
        pos++;
        final e = s.codeUnitAt(pos);
        if (e == 110) { buf!.write('\n'); } else if (e == 116) { buf!.write('\t'); }
        else if (e == 117) { buf!.write(String.fromCharCode(hex4())); }
        else { buf!.write(String.fromCharCode(e)); }
        pos++;
        start = pos;
        continue;
      }
      pos++;
    }
    final tail = s.substring(start, pos);
    pos++;
    if (buf == null) return tail;
    buf.write(tail);
    return buf.toString();
  }

  int hex4() {
    var v = 0;
    for (var i = 0; i < 4; i++) {
      final c = s.codeUnitAt(pos);
      var d = c - 48;
      if (d > 9) d = (c | 32) - 87;
      v = v * 16 + d;
      pos++;
    }
    return v;
  }

  num number() {
    var neg = false;
    if (s.codeUnitAt(pos) == 45) { neg = true; pos++; }
    var n = 0;
    while (pos < len) {
      final c = s.codeUnitAt(pos);
      if (c < 48 || c > 57) break;
      n = n * 10 + c - 48;
      pos++;
    }
    if (pos < len && s.codeUnitAt(pos) == 46) {
      pos++;
      var scale = 0.1;
      var frac = 0.0;
      while (pos < len) {
        final c = s.codeUnitAt(pos);
        if (c < 48 || c > 57) break;
        frac += (c - 48) * scale;
        scale *= 0.1;
        pos++;
      }
      final v = n + frac;
      return neg ? -v : v;
    }
    return neg ? -n : n;
  }
}

void writeString(String v, StringBuffer b) {
  b.write('"');
  var start = 0;
  for (var i = 0; i < v.length; i++) {
    final c = v.codeUnitAt(i);
    if (c == 34 || c == 92) {
      if (i > start) { b.write(v.substring(start, i)); }
      b.write('\\');
      b.writeCharCode(c);
      start = i + 1;
    } else if (c == 10) {
      if (i > start) { b.write(v.substring(start, i)); }
      b.write('\\n');
      start = i + 1;
    }
  }
  if (v.length > start) { b.write(v.substring(start)); }
  b.write('"');
}

void writeValue(Object? v, StringBuffer b) {
  if (v == null) { b.write('null'); return; }
  if (v is bool) { b.write(v ? 'true' : 'false'); return; }
  if (v is num) { b.write(v); return; }
  if (v is String) { writeString(v, b); return; }
  if (v is List) {
    b.write('[');
    for (var i = 0; i < v.length; i++) {
      if (i > 0) { b.write(','); }
      writeValue(v[i], b);
    }
    b.write(']');
    return;
  }
  final m = v as Map;
  b.write('{');
  var first = true;
  m.forEach((key, val) {
    if (!first) { b.write(','); }
    first = false;
    writeString(key as String, b);
    b.write(':');
    writeValue(val, b);
  });
  b.write('}');
}

String makeDoc(int seed) {
  var x = seed;
  int next() { x = (x * 1103515245 + 12345) & 0x7fffffff; return x; }
  final b = StringBuffer();
  b.write('{"items":[');
  for (var i = 0; i < 40; i++) {
    if (i > 0) b.write(',');
    b.write('{"id":');
    b.write(next() % 100000);
    b.write(',"name":"user_');
    b.write(next() % 9999);
    b.write('","tags":["alpha","b');
    b.write(next() % 999);
    b.write('","g"],"score":');
    b.write((next() % 10000) / 100);
    b.write(',"ok":');
    b.write(next() % 3 == 0 ? 'true' : 'false');
    b.write(',"meta":null}');
  }
  b.write('],"count":40}');
  return b.toString();
}

int checksum(Object? v) {
  if (v is List) {
    var n = v.length;
    for (final e in v) { n += checksum(e); }
    return n;
  }
  if (v is Map) {
    var n = v.length;
    v.forEach((k, e) { n += k.length + checksum(e); });
    return n;
  }
  if (v is String) return v.length;
  if (v is int) return v & 0xff;
  if (v is double) return v.toInt() & 0xff;
  return v == null ? 0 : 1;
}

int main(int n) {
  var total = 0;
  for (var i = 0; i < n; i++) {
    final doc = makeDoc(123456789 + i * 2654435761);
    final parsed = Parser(doc).value();
    final out = StringBuffer();
    writeValue(parsed, out);
    total += out.length + checksum(parsed);
  }
  return total;
}
''';

void main(List<String> args) {
  final iterations = args.isEmpty ? 4000 : int.parse(args[0]);
  final samples = args.length < 2 ? 7 : int.parse(args[1]);
  final compiler = Compiler();
  compiler.entrypoints.add('package:json_codec/main.dart');
  final program = compiler.compile({
    'json_codec': {'main.dart': _source},
  });
  final runtime = Runtime(program.write().buffer);
  int run(int n) =>
      runtime.executeLib(
        'package:json_codec/main.dart',
        'main',
        arguments: {'n': n},
      )
          as int;
  var checksum = 0;
  for (var warm = 0; warm < 2; warm++) {
    checksum += run(50);
  }
  final times = <double>[];
  for (var sample = 0; sample < samples; sample++) {
    final watch = Stopwatch()..start();
    final result = run(iterations);
    watch.stop();
    checksum += result;
    times.add(watch.elapsedMicroseconds / 1000);
  }
  final raw = List<double>.of(times);
  times.sort();
  final median = times[times.length ~/ 2];
  print(
    'json_codec median_ms=${median.toStringAsFixed(3)} '
    'min_ms=${times.first.toStringAsFixed(3)} '
    'max_ms=${times.last.toStringAsFixed(3)} '
    'ns/iteration=${(median * 1000000 / iterations).toStringAsFixed(2)} '
    'raw_ms=${raw.map((value) => value.toStringAsFixed(3)).join(',')}',
  );
  print('checksum=$checksum');
}
